woocommerce/woocommerce
# Use MC Stats for PHP fatal error counting

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Part of https://github.com/woocommerce/woocommerce/issues/49289

This PR creates a McStats class which extends from [Jetpack's `A8c_Mc_Stats`](https://github.com/Automattic/jetpack-a8c-mc-stats/blob/trunk/tests/php/test_Stats.php) class for recording stats, and use it to record fatal errors.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site using this branch with code snippet plugin installed and activated.
2. Go to Core Profiler and allow tracking
3. Add the following code snippet to the code snippet plugin:

```php
function custom_wp_remote_get_debug( $response, $args, $url ) {
    if ( strpos( $url, 'https://pixel.wp.com' ) === 0 ) {
        // Print the request URL
        echo 'Request URL: ' . esc_url( $url ) . '<br>';

        // Check if the response is a WP_Error or not
        if ( is_wp_error( $response ) ) {
            echo 'Response: Failed<br>';
        } else {
            echo 'Response: Success<br>';
        }
    }
    return $response;
}

add_filter( 'http_response', 'custom_wp_remote_get_debug', 10, 3);
```

4. Trigger a fatal error in WooCommerce by adding the following code to [load_rest_api()](https://github.com/woocommerce/woocommerce/blob/ac834f55f75d4e51955bad43115939faa38ec9da/plugins/woocommerce/includes/class-woocommerce.php#L535):

```php
throw new Exception( 'Test fatal error in woocommerce' );
```
5. Reload the page
6. Confirm that the screen shows the `Response: Success` 

![Screenshot 2024-07-24 at 10 59 17](https://github.com/user-attachments/assets/77eda2d6-c067-4ec7-a990-6d2e9234239c)






### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
